src/mgmt/cli_mgmt.c

```diff
@@ -1271,6 +1272,158 @@ dc_mgmt_tm_register(const char *sys, const char *jobid, key_t shm_key, uid_t *ow
+
+	ms_ranks = sys->sy_info.ms_ranks;
+	D_ASSERT(ms_ranks->rl_nr > 0);
+
+	rc = rsvc_client_init(&ms_client, ms_ranks);
+	if (rc != 0) {
+		DL_ERROR(rc, "failed to init ms client");
+		D_GOTO(out_grp, rc);
+	}
+
+	ep.ep_grp = sys->sy_group;
+	ep.ep_tag = daos_rpc_tag(DAOS_REQ_MGMT, 0);
+	opc       = DAOS_RPC_OPCODE(MGMT_POOL_LIST, DAOS_MGMT_MODULE, DAOS_MGMT_VERSION);
+
+rechoose:
+	rc = rsvc_client_choose(&ms_client, &ep);
+	if (rc != 0) {
+		DL_ERROR(rc, "failed to choose MS rank");
+		D_GOTO(out_client, rc);
+	}
+
+	rc = crt_req_create(daos_task2ctx(task), &ep, opc, &rpc);
+	if (rc != 0) {
+		DL_ERROR(rc, "crt_req_create(MGMT_POOL_LIST) failed");
+		D_GOTO(out_client, rc);
+	}
+
+	in          = crt_req_get(rpc);
+	in->pli_grp = (d_string_t)args->grp;
+	/* If provided pools is NULL, caller needs the number of pools
+	 * to be returned in npools. Set npools=0 in the request in this case
+	 * (caller value may be uninitialized).
+	 */
+	if (args->pools == NULL)
+		in->pli_npools = 0;
+	else
+		in->pli_npools = *args->npools;
+
+	/* Now fill in the client credential for the pool access checks. */
+	rc = dc_sec_request_creds(&in->pli_cred);
+	if (rc != 0) {
+		DL_ERROR(rc, "failed to obtain security credential");
+		D_GOTO(out_put_req, rc);
+	}
+
+	D_DEBUG(DB_MGMT, "req_npools=" DF_U64 " (pools=%p, *npools=" DF_U64 "\n", in->pli_npools,
+		args->pools, *args->npools);
+
+	crt_req_addref(rpc);
+	rc = daos_rpc_send_wait(rpc);
+	if (rc != 0) {
+		DL_ERROR(rc, "rpc send failed");
+		crt_req_decref(rpc);
+		goto rechoose;
+	}
+
+	out = crt_reply_get(rpc);
+	D_ASSERT(out != NULL);
+
+	rc = rsvc_client_complete_rpc(&ms_client, &ep, rc, out->plo_op.mo_rc, &out->plo_op.mo_hint);
+	if (rc == RSVC_CLIENT_RECHOOSE) {
+		crt_req_decref(rpc);
+		goto rechoose;
+	}
+
+	rc = out->plo_op.mo_rc;
+	if (rc != 0)
+		D_GOTO(out_put_req, rc);
+
+	*args->npools = out->plo_npools;
+
+	/* copy RPC response pools info to client buffer, if provided */
+	if (args->pools) {
+		/* Response ca_count expected <= client-specified npools */
+		for (pidx = 0; pidx < out->plo_pools.ca_count; pidx++) {
+			struct mgmt_pool_list_pool *rpc_pool = &out->plo_pools.ca_arrays[pidx];
+			daos_mgmt_pool_info_t      *cli_pool = &args->pools[pidx];
+
+			uuid_copy(cli_pool->mgpi_uuid, rpc_pool->plp_uuid);
+
+			D_STRNDUP(cli_pool->mgpi_label, rpc_pool->plp_label,
+				  DAOS_PROP_LABEL_MAX_LEN);
+			if (cli_pool->mgpi_label == NULL) {
+				D_ERROR("copy RPC reply label failed\n");
+				D_GOTO(out_free_args_pools, rc = -DER_NOMEM);
+			}
+
+			/* allocate rank list for caller (simplifies API) */
+			rc = d_rank_list_dup(&cli_pool->mgpi_svc, rpc_pool->plp_svc_list);
+			if (rc != 0) {
+				D_ERROR("copy RPC reply svc list failed\n");
+				D_GOTO(out_free_args_pools, rc = -DER_NOMEM);
+			}
+		}
+	}
+
+out_free_args_pools:
+	if (args->pools && (rc != 0)) {
+		for (pidx = 0; pidx < out->plo_pools.ca_count; pidx++) {
+			daos_mgmt_pool_info_t *pool = &args->pools[pidx];
+
+			if (pool->mgpi_label)
```

Comment:
Is it sure that mgpi_lable and mgpi_svc are initialized to zero ?

Reply:
Hmm. I guess not, because the caller is responsible for allocating that array so we don't know for sure what they did. I added explicit NULL assignments prior to copying from the RPC response in the loop.